pennylane/beta/tapes/qubit_param_shift.py

```diff
@@ -0,0 +1,228 @@
+            self.analytic_pd = self.parameter_shift_var
+
+            # Make a copy of the original measurements; we will be mutating them
+            # during the parameter shift method.
+            self._original_measurements = self._measurements.copy()
+
+            # Finally, store the locations of any variance measurements in the
+            # measurement queue.
+            self.var_idx = np.where(self.var_mask)[0]
+
+    def _grad_method(self, idx, use_graph=True, default_method="A"):
+        op = self._par_info[idx]["op"]
+
+        if op.grad_method == "F":
+            return "F"
+
+        return super()._grad_method(idx, use_graph=use_graph, default_method=default_method)
+
+    def jacobian(self, device, params=None, **options):
+        # The parameter_shift_var method needs to evaluate the circuit
+        # at the unshifted parameter values; these are stored in the
+        # self._evA attribute. Here, we set the value of the attribute to None
+        # before each Jacobian call, so that the expectation value is calculated only once.
+        self._evA = None
+        return super().jacobian(device, params, **options)
+
+    def parameter_shift(self, idx, device, params, **options):
+        r"""Partial derivative using the parameter-shift rule of a tape consisting of measurement
+        statistics that can be represented as expectation values of observables.
+
+        This includes tapes that output probabilities, since the probability of measuring a
+        basis state :math:`|i\rangle` can be written in the form of an expectation value:
+
+        .. math::
+
+            \mathbb{P}_{|i\rangle} = |\langle i | U(\mathbf(p)) | 0 \rangle|^2
+                = \langle 0 | U(\mathbf(p))^\dagger | i \rangle\langle i | U(\mathbf(p)) | 0 \rangle
+                = \mathbb{E}\left( | i \rangle\langle i | \right)
+
+        Args:
+            idx (int): trainable parameter index to differentiate with respect to
+            device (.Device, .QubitDevice): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): the quantum tape operation parameters
+
+        Keyword Args:
+            shift (float): the parameter shift value
+
+        Returns:
+            array[float]: 1-dimensional array of length determined by the tape output
+            measurement statistics
+        """
+        t_idx = list(self.trainable_params)[idx]
+        op = self._par_info[t_idx]["op"]
+        p_idx = self._par_info[t_idx]["p_idx"]
+
+        s = (
+            np.pi / 2
+            if op.grad_recipe is None or op.grad_recipe[p_idx] is None
+            else op.grad_recipe[p_idx]
+        )
+        s = options.get("shift", s)
+
+        shift = np.zeros_like(params)
+        shift[idx] = s
+
+        shift_forward = np.array(self.execute_device(params + shift, device))
+        shift_backward = np.array(self.execute_device(params - shift, device))

```

Comment:
Just to think into the future...we would soon change this to be agnostic towards the number of shifts and do something like:

``` python
shift_data = options.get("shift_data")

shifted = [coeff*np.array(self.execute_device(params + shift, device)) for coeff, shift in shift_data] 

return np.sum(shifted)

```

Reply:
oh yep. exactly what I was thinking!